### PR TITLE
(I hate titles)

### PR DESCRIPTION
**Pull request by wecing**
TL;DR: I didn't add the exception feature we talked about...

...because pyodbc will throw a ProgrammingError for us; the error message is a little bit confusing("error converting numeric to varchar") but trying to catch the exception and change the message might introduce some other problems.
(yes, 576 additions and 66 deletions -- don't worry; one half of them are tests.)

### CODE REVIEW

File: src/cursor.cpp [hunk 1]
```diff
@@ -840,12 +856,30 @@ static PyObject* Cursor_callproc(PyObject* self, PyObject* args)
 
     PyObject* pReturn = 0;
 
+    // FIXME: assuming the method name does not contain characters beyond (extended?) ascii.
+    //
+    // Under Python 3, PyString_AsString is a macro provided by pyodbc which expands to PyUnicode_AsString,
+    // which is obsolete from Py 3.3(?). So maybe we should remove the macro def in pyodbccompat.h.
+#if PY_VERSION_HEX >= 0x03000000
+    PyObject* pCallprocName = PyUnicode_AsASCIIString(pProcName);
```
Comment (joshuahlang) [inline src/cursor.cpp:864]: Check pCallprocName for NULL

File: src/cursor.cpp [hunk 2]
```diff
@@ -810,14 +810,30 @@ static PyObject* Cursor_callproc(PyObject* self, PyObject* args)
         return 0;
     }
 
-    PyObject* pProcName = PyTuple_GET_ITEM(args, 0);
+    free_results(cursor, FREE_STATEMENT | KEEP_PREPARED);
+    free_results(cursor, FREE_STATEMENT | KEEP_PREPARED); // FIXME: why?
```
Comment (joshuahlang) [inline src/cursor.cpp:814]: Did you ever determine why this was necessary?

File: src/params.cpp [hunk 3]
```diff
@@ -184,29 +196,69 @@ static bool GetBytesInfo(Cursor* cur, Py_ssize_t index, PyObject* param, ParamIn
     info.ValueType = SQL_C_BINARY;
     info.ColumnSize = (SQLUINTEGER)max(len, 1);
 
+    // FIXME: share code between py3 and py2
```
Comment (joshuahlang) [inline src/params.cpp]: Should cleanup fixmes

File: src/params.cpp [hunk 4]
```diff
@@ -184,29 +196,69 @@ static bool GetBytesInfo(Cursor* cur, Py_ssize_t index, PyObject* param, ParamIn
...
     if (len <= cur->cnxn->varchar_maxlength)
     {
-        info.ParameterType     = SQL_VARCHAR;
-        info.StrLen_or_Ind     = len;
-        info.ParameterValuePtr = PyBytes_AS_STRING(param);
+        info.ParameterType = SQL_VARCHAR;
+        info.StrLen_or_Ind = len;
+        // FIXME: INPUT_STREAM & INPUT_OUTPUT_STREAM are not supported in unixODBC for now.
+        if (info.InputOutputType == SQL_PARAM_INPUT) {
+            info.BufferLength = len + 1; // should not be used?
```
Comment (joshuahlang) [inline src/params.cpp]: What does this comment mean? Is this adding a byte for the '\0' terminator?

File: src/params.cpp [hunk 5]
```diff
@@ -232,41 +287,53 @@ static bool GetUnicodeInfo(Cursor* cur, Py_ssize_t index, PyObject* param, Param
...
-        info.StrLen_or_Ind = (SQLINTEGER)(len * sizeof(SQLWCHAR));
     }
     else
     {
         // Too long to pass all at once, so we'll provide the data at execute.
-
         info.ParameterType     = SQL_WLONGVARCHAR;
         info.StrLen_or_Ind     = cur->cnxn->need_long_data_len ? SQL_LEN_DATA_AT_EXEC((SQLLEN)len * sizeof(SQLWCHAR)) : SQL_DATA_AT_EXEC;
         info.ParameterValuePtr = param;
+        // FIXME: fnToPyObject?
```
Comment (joshuahlang) [inline src/params.cpp:293]: Can this be fixed?

File: src/params.cpp [hunk 6]
```diff
@@ -184,29 +196,69 @@ static bool GetBytesInfo(Cursor* cur, Py_ssize_t index, PyObject* param, ParamIn
...
+        info.StrLen_or_Ind = len;
+        // FIXME: INPUT_STREAM & INPUT_OUTPUT_STREAM are not supported in unixODBC for now.
+        if (info.InputOutputType == SQL_PARAM_INPUT) {
+            info.BufferLength = len + 1; // should not be used?
+            info.ParameterValuePtr = PyBytes_AS_STRING(param);
+        } else {
+            if (ostr_len < len) {
+                ostr_len = (int)len;
+            }
+            void* buf = malloc(ostr_len + 1);
```
Comment (joshuahlang) [inline src/params.cpp]: Should this be pyodbc_malloc?

File: src/params.cpp [hunk 7]
```diff
@@ -546,10 +664,28 @@ static bool GetBufferInfo(Cursor* cur, Py_ssize_t index, PyObject* param, ParamI
...
         info.StrLen_or_Ind     = cb;
+
+        if (info.InputOutputType == SQL_PARAM_INPUT) {
+            info.ParameterValuePtr = (SQLPOINTER)pb;
+            info.BufferLength      = cb;
+        } else {
+            if (obuf_len < cb) {
+                obuf_len = (int)cb;
+            }
+            void* buf = malloc(obuf_len);
```
Comment (joshuahlang) [inline src/params.cpp]: pyodbc_malloc?

File: src/params.cpp [hunk 8]
```diff
@@ -564,23 +700,45 @@ static bool GetBufferInfo(Cursor* cur, Py_ssize_t index, PyObject* param, ParamI
...
+        info.ColumnSize = (SQLUINTEGER)max(cb, 1);
+        info.StrLen_or_Ind = cb;
+        if (info.InputOutputType == SQL_PARAM_INPUT) {
+            info.BufferLength = cb;
+            info.ParameterValuePtr = PyByteArray_AS_STRING(param);
+        } else {
+            if (obuf_len < cb) {
+                obuf_len = (int)cb;
+            }
+            void* buf = malloc(obuf_len);
```
Comment (joshuahlang) [inline src/params.cpp]: pyodbc_malloc?

File: tests2/freetdstests.py [hunk 9]
```diff
@@ -1229,6 +1229,158 @@ def test_no_fetch(self):
...
+    #
+
+    def test_callproc_output_varchar_int(self):
+        self.cursor.execute('''
+                            create procedure proc1
+                                @a varchar(max) OUTPUT,
+                                @b int output
+                            as
+                            begin
+                                select @a = (select @@VERSION),
```
Comment (joshuahlang) [inline tests2/freetdstests.py]: I would use a static string here instead of @@VERSION so you're sure what is coming out of r the stored procedure regardless of the underlying database

File: src/cursor.cpp (hunk 2 above)
Comment (wecing) [inline src/cursor.cpp:814]: I tried. I traced the control flow in mysql connector. But I still don't understand know why I need to call it twice. There are no asynchronous calls inside the connector. On the first time it's called it will set one flag from 1 to 0, and the second time, because of that, the program will fall onto a different if-else branch. On the first time it's executed it would call mysql_stmt_get_results(); and the second time it would call mysql_get_results() instead. But according to the spec the second call should be identical to the first.

In short, I don't understand why.

File: src/cursor.cpp (hunk 1 above)
Comment (wecing) [inline src/cursor.cpp:864]: done.

File: src/params.cpp (hunk 3 above)
Comment (wecing) [inline src/params.cpp]: done.

File: src/params.cpp (hunk 5 above)
Comment (wecing) [inline src/params.cpp:293]: I don't know how to retrieve output when input uses data-at-execution.

File: src/params.cpp (hunk 6 above)
Comment (wecing) [inline src/params.cpp]: all malloc()'s are changed to pyodbc_malloc().
